File: matrix_multiplication_order.f90
program main
    implicit none
    
    integer N
    real*8, dimension(:),   allocatable :: vector, result, resultJI
    real*8, dimension(:,:), allocatable :: matrix
    real start, finish, time_taken
    call random_seed()
    
    open(1, file = './results/fortran_IJ.csv', action = 'write')
    open(2, file = './results/fortran_JI.csv', action = 'write')
    
    do N = 1000, 33000, 1000

        allocate(vector(N))
        allocate(matrix(N, N))
    
        call randomVector(vector, N)
        call randomSquareMatrix(matrix, N)
    
        allocate(result(N))
        call emptyVector(result, N)
    
        call cpu_time(start)
        call matrixVectorMultiplyij(matrix, vector, result, N)
        call cpu_time(finish)
    
        time_taken =  finish - start
        print *, 'For size ', N, 'the matrix multiplication IJ took ', time_taken, ' second(s) to execute'
        write(1,*) N,",",time_taken
        deallocate(result)
    
        allocate(resultJI(N))
        call emptyVector(resultJI,N)
    
        call cpu_time(start )
        call matrixVectorMultiplyji(matrix,vector,resultJI,N)
        call cpu_time(finish )
    
        time_taken  =  finish  - start
        print *, 'For size ', N, 'the matrix multiplication JI took ', time_taken, ' second(s) to execute'
        write(2,*) N,",",time_taken
        deallocate(resultJI)
    
        deallocate(matrix)
        deallocate(vector)
    end do
    
    close(1)
    close(2)
    
    contains
    
    subroutine randomVector(vector, N)
        implicit none

        real(8), dimension(:) :: vector
        integer :: N, i
        real(8) :: number
        do i = 1, N
            call random_number(number)
            vector(i) = number * (N + 1)
        end do 
    end
    
    subroutine emptyVector(vector, N)
        implicit none

        real(8), dimension(:) :: vector
        integer :: N, i
        do i = 1, N
            vector(i) = 0
        end do 
    end
    
    subroutine randomSquareMatrix(matrix, N)
        implicit none

        real(8), dimension(:,:) :: matrix
        integer :: N, i, j
        real(8) :: number

        do i = 1, N
            do j = 1, N
                call random_number(number)
                matrix(i, j) = number * (N + 1)
            end do
        end do 
    end
    
    subroutine matrixVectorMultiplyij(matrix, vector, result, N)
        implicit none
    
        real*8, dimension(:,:) :: matrix
        real*8, dimension(:) :: vector, result
        integer :: N, i, j
    
        do i = 1, N
            do j = 1, N
                result(i) = result(i) + (matrix(i, j) * vector(j))
            end do
        end do
    end
    
    subroutine matrixVectorMultiplyji(matrix, vector, result, N)
        implicit none
    
        real*8, dimension(:,:) :: matrix
        real*8, dimension(:) :: vector, result
        integer :: N, i, j
    
        do j = 1, N
            do i = 1, N
                result(i) = result(i) + (matrix(i, j) * vector(j))
            end do
        end do
    end
    
    end program main

File: matrix_multiplication_order.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include <stdbool.h>
#include <string.h>
#include <ctype.h>

double randomNumber() {
    return (double)rand() / (double)RAND_MAX;
}

double **randomSquareMatrix(int N) {
    double **matrix = (double **)malloc(N * sizeof(double *));
    for (int i = 0; i < N; i++) {
        matrix[i] = (double *)malloc(N * sizeof(double));   
        for (int j = 0; j < N; j++) matrix[i][j] = randomNumber();
    };
    return matrix;
}

double *randomVector(int N) {
    double *vector = (double *)malloc(N * sizeof(double));
    for (int i = 0; i <= N; i++) vector[i] = 0;

    return vector;
}


double *emptyVector(int N) {
    double *vector = (double *)malloc(N * sizeof(double));
    for (int i = 0; i <= N; i++) vector[i] = 0;

    return vector;
}

double *matrixVectorMultiplyij(double **matrix, double *vector, int N) {
    double *result = emptyVector(N);

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) result[i] += matrix[i][j] * vector[j];
    }
    return result;
}

double *matrixVectorMultiplyji(double **matrix, double *vector, int N) {
    double *result = emptyVector(N);

    for (int j = 0; j < N; j++) {
        for (int i = 0; i < N; i++) result[i] += matrix[i][j] * vector[j];
    }
    return result;
}


int main() {

    FILE *ij, *ji;

    ij = fopen("./results/c_IJ.csv", "w+");
    ji = fopen("./results/c_JI.csv", "w+");


    for (int i = 1; i <= 33; i++){
        int N = 1000 * i;
    
        // create the matrix and vectors and fill them with random numbers or empty
        double **matrix = randomSquareMatrix(N);
        double *vector  = randomVector(N);
        double *result;

        // multiply the matrix and vector print the time taNen

        // IJ
        clock_t start, finish;
        start = clock();

        result = matrixVectorMultiplyij(matrix, vector, N);

        finish = clock();
        
        
        double time_taken = ((double)(finish - start)) / CLOCKS_PER_SEC;;  // in seconds

        printf("For size %d, the matrix multiplication IJ took %.6f second(s) to execute \n", N, time_taken);
        fflush(ij);
        fprintf(ij, "%d, %.6f\n", N, time_taken);


        // JI
        start = clock();

        result = matrixVectorMultiplyji(matrix, vector, N);

        finish = clock();

        time_taken = ((double)(finish - start)) / CLOCKS_PER_SEC;;  // in seconds

        printf("For size %d, the matrix multiplication JI took %.6f second(s) to execute \n", N, time_taken);
        fprintf(ji, "%d, %.6f\n", N, time_taken);
    }
    fclose(ij);
    fclose(ji);

    

    return 0;
}
